ipp/src/parser.rs

```diff
@@ -189,6 +188,22 @@ where
             }
         }
 
+        Ok(header)
+    }
+
+    /// Parse IPP stream without reading beyond the end of the attributes. The payload stays untouched.
+    pub async fn parse_without_payload(mut self) -> Result<(IppHeader, IppAttributes), IppParseError> {
```

Comment:
I am wondering if consuming `self` makes sense here, originally it was for eventual payload consumption. May be we could say `&mut self`, which will give an additional advantage of using the parser again with something like `into_payload()`.
Another possibility: keep `mut self`, but return `PartialIppResponse` instance with header, attributes and the reader object.

Reply:
Hmmm you're right of course. I wonder if there is a better course of action here.

How about using a [typestate pattern](https://docs.rust-embedded.org/book/static-guarantees/typestate-programming.html)? We could have something like `IppParser<State>` which could be `IppParser<Initial>`, `IppParser<Attributes>` etc? We basically have 3 states: initial, after header & before attributes, after attributes.

Or would this be overkill for such a simple state machine? Also, this would obviously be a breaking change.

Reply:
Not sure introducing generics would be much beneficial.
I can also suggest this:

```rust
pub async fn parse_attributes(&mut self) -> Result<(IppHeader, &IppAttributes), IppParseError> {
    let header = self.parse_header_attributes().await?;
    Ok((header, &self.state.attributes))
}

pub fn reader(&mut self) -> &mut R {
    &self.reader
}

```

Reply:
The disadvantage of both of those is that the user only gets a `&IppAttributes` (which is stored in `state.attributes`), and not `IppAttributes` which is usually what a user wants after invoking a parser. 

How about, instead of `parse_without_payload`, we will have
```rust
/// Parse IPP stream without reading beyond the end of the attributes. The payload stays untouched.
pub async fn parse_parts(mut self) -> Result<(IppHeader, IppAttributes, AsyncIppReader<R>), IppParseError> {
    let header = self.parse_header_attributes().await?;
    Ok((header, self.state.attributes, self.reader))
}
```

That way, the function will work as a sort of parallel to `parse` except that it does not wrap the remaining data into an `IppPayload` (which might be undesired as explained above due to `'static` requirements and disadvantages of `Box<dyn ...>`)

Reply:
Looks ok to me.

Reply:
Pushed an update!